### DISCUSSION TITLE
Are uploaded file URLs intended to be stable? (i.e. can they be stored in the database)

### DISCUSSION THREAD

**Discussion by lukehutch** [serverpod/serverpod#994] [Q&A] [answered]
When you want to fetch a file from the Serverpod server, from a file that has been uploaded, the server has to look up the URL for the file, based on its upload path, by calling `Future<Uri?> session.storage.getPublicUrl(storageId: String, path: String)`.

Assuming the hostname and port of the server don't change, is there any chance the URL of a file will change for some reason? Or is it safe to assume file URLs are stable?

The reason I ask is because it is inefficient to make server roundtrips for a large set of file paths whose URLs you need in the client. I added #922 to mitigate this (you pass in a list of file paths, and you get back a list of URLs, in a single call with maximum concurrency), by calling `Future<List<Uri?>> session.storage.getPublicUrls(storageId: String, paths: List<String>)`.

However, it would be much more efficient to look up the file URL when it is uploaded to the server, and to simply store the file URL in the database rather than the file path. That way, any database result set already has the URL looked up for any image resources it refers to.

Would it be a bad idea to store file URLs in the database?

**Comment by vlidholt** [comment, marked as answer]
File URLs can be considered stable as long as you keep your setup the same. E.g. if you use a bucket with a specific domain name, it will stay the same. If you change from database storage to cloud storage it may change.
